Objects/unionobject.c

```diff
@@ -237,20 +237,24 @@ dedup_and_flatten_args(PyObject* args)
         PyObject* i_element = PyTuple_GET_ITEM(args, i);
         for (Py_ssize_t j = i + 1; j < arg_length; j++) {
             PyObject* j_element = PyTuple_GET_ITEM(args, j);
+            int is_ga = Py_TYPE(i_element) == &Py_GenericAliasType &&
+                        Py_TYPE(j_element) == &Py_GenericAliasType;
             // RichCompare to also deduplicate GenericAlias types (slower)
-            if (Py_TYPE(i_element) == &Py_GenericAliasType &&
-                Py_TYPE(j_element) == &Py_GenericAliasType) 
-            {
-                if (PyObject_RichCompareBool(i_element, j_element, Py_EQ) == 1) {
-                    is_duplicate = 1;
-                }
+            is_duplicate = is_ga ? PyObject_RichCompareBool(i_element, j_element, Py_EQ)
+                : i_element == j_element;
+            // Should only happen if RichCompare fails
+            if (is_duplicate < 0) {
+                PyErr_Format(PyExc_TypeError, 
+                    "Could not compare objects of type '%s' and '%s'"
+                    " at indexes %d and %d respectively."
+                    " Their __eq__ methods may be invalid. ",
+                    Py_TYPE(i_element)->tp_name, Py_TYPE(j_element)->tp_name,
+                    i, j);
+                Py_DECREF(args);
+                return NULL;
```

Comment:
I'm really sorry, but this is way more than you need. All I had meant with my original comment was something like this:
```
is_duplicate = PyObject_RichCompareBool(i_element, j_element, Py_EQ);
if (is_duplicate < 0) {
    Py_DECREF(args);
    return NULL;
}
```
This will pass the error back up. The problem with calling `PyErr_format()` is that it overwrites whatever error was raised previously.

A test case to see for yourself would be:
```py
class BadType(type):
    def __eq__(self, other):
        1/0
bt = BadType('bt', (), {})
list[int] | list[bt]
```
Your version masks the ZeroDivisionError; if you just remove the `PyErr_Format()` call, it shows the right traceback.

Reply:
Thanks, and sorry this took so many reviews for a relatively minor change.